[FEATURE] Document Model and Repository (#6531)
TYPO3-Documentation/TYPO3CMS-Reference-CoreApi: Documentation/ExtensionArchitecture/ExtbaseRewrite/Appendix/Attributes.rst

:navigation-title: PHP attributes reference

..  include:: /Includes.rst.txt
..  index:: pair: Extbase; PHP attributes
..  _extbase-appendix-attributes:

================================
Extbase PHP attributes reference
================================

..  include:: /ExtensionArchitecture/ExtbaseRewrite/_wip.rst.txt

This page is a quick-reference for all PHP attributes defined and handled by

> the Extbase framework itself. It covers ORM attributes (used on model

Reviewer: ```suggestion
the Extbase framework itself. It covers :abbr:`ORM (Object Relational Mapping)` attributes (used on model
```

For abbreviations like this, I would always use the :abbr: role on first occurrence in a section , as junior developers may not be familiar with them. This also enables users to search for both terms

Author: done

properties), validation attributes (used on model properties and action
parameters), and controller-level attributes (used on action methods).

It does not cover attributes defined by TYPO3 Core outside of Extbase, or
PHP built-in attributes. Where a more detailed explanation exists in another
chapter, this page links there.

..  contents:: On this page
    :local:

>     :depth: 1

Reviewer: I would suggest to use more then one level, enabling experienced users to jump to the attribute they want to read on right now

Author: done

..  _extbase-appendix-attributes-orm:

> ORM attributes (persistence)

Reviewer: ```suggestion
ORM  (Object Relational Mapping) attributes (persistence)
```

This elevates the section in search results and again helps new devs orientate

Author: done

=============================

These attributes are declared in the :php:`\TYPO3\CMS\Extbase\Attribute\ORM`
namespace and are placed on
model properties to control how Extbase maps them to and from the database.

..  _extbase-appendix-attributes-lazy:

#[Lazy]
-------

:Target: Model property
:Namespace: :php:`\TYPO3\CMS\Extbase\Attribute\ORM\Lazy`
:Parameters: none

Defers loading of a related object or :php:`ObjectStorage` until the property
is first accessed. Without :php:`#[Lazy]`, Extbase loads all related objects
when the parent is loaded — which can trigger N+1 queries in list views.

The proxy resolves itself automatically via PHP magic methods on any property
or method access — you do not need to call :php:`_loadRealInstance()` yourself.

When applied to a 1:1 relation, the property type must include

> :php:`LazyLoadingProxy` so Extbase knows to install the proxy. A typed getter

Reviewer: ```suggestion
:php-short:`\TYPO3\CMS\Extbase\Persistence\Generic\LazyLoadingProxy` so Extbase knows to install the proxy. A typed getter
```

Use the php-short role so that a help tooltip for the class can be displayed and ppl can jump to api.typo3.org or the source code on GitHub. using :php-short: instead of :php: only the short name is displayed

Author: done

needs an :php:`instanceof` check only so PHPStan and your IDE can narrow the
return type to :php:`?Location`:

..  code-block:: php
    :caption: EXT:my_extension/Classes/Domain/Model/Event.php

    use TYPO3\CMS\Extbase\Attribute\ORM\Lazy;
    use TYPO3\CMS\Extbase\Persistence\Generic\LazyLoadingProxy;

    #[Lazy]

>     protected Location|LazyLoadingProxy|null $location = null;

Reviewer: I would wrap this in a class, so that the context (This is a model not a controller) becomes visble at a glance and new users can more easily be orientated

Author: done

..  _extbase-appendix-attributes-ignorevalidation:

#[IgnoreValidation]
-------------------

:Target: Action method parameter
:Namespace: :php:`\TYPO3\CMS\Extbase\Attribute\IgnoreValidation`
:Parameters: none (place directly on the parameter)

Suppresses validation for one action parameter. Useful in multi-step forms
where partial state is forwarded between actions:

..  code-block:: php
    :caption: EXT:my_extension/Classes/Controller/EventController.php

    use TYPO3\CMS\Extbase\Attribute\IgnoreValidation;

    public function previewAction(
        #[IgnoreValidation] Event $event
    ): ResponseInterface {
        // validation skipped for $event
    }

..  note::

    In TYPO3 v14, :php:`#[IgnoreValidation]` must be placed on the
    **parameter**, not on the method with an :php:`argumentName` property.
    The method-level form was deprecated in v14 (Deprecation `#108227
    <https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/14.0/Deprecation-108227-UsageOfIgnoreValidationAndValidateAttributesForParametersAtMethodLevel.html>`__).

..  _extbase-appendix-attributes-controller:

> Controller attributes (v14)

Reviewer: Avoid referring to versions in headlines, use the versionchanged or versionadded directives instead:

[Versions](https://docs.typo3.org/permalink/h2document:version-hints)

As we can automatically detect and remove such version hints in later Versions. Also all this content is only backported into v14 and v15 anyway

Author: done

============================

These attributes are declared in the :php:`\TYPO3\CMS\Extbase\Attribute\`
namespace and are applied to controller action methods to control access and

> rate limiting. Both are new in TYPO3 v14.

Reviewer: when something was new should always go in the [Versionadded](https://docs.typo3.org/m/typo3/docs-how-to-document/main/en-us/Reference/ReStructuredText/Content/Versions.html#versionadded) directive

Author: done

Full documentation for these attributes is planned for the Controller and
Security chapters.

..  _extbase-appendix-attributes-authorize:

#[Authorize]
------------

:Target: Controller action method
:Namespace: :php:`\TYPO3\CMS\Extbase\Attribute\Authorize`

> :New in: TYPO3 v14 (Feature `#107826 <https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/14.2/Feature-107826-IntroduceExtbaseActionAuthorizationLogic.html>`__)

Reviewer: use the versionadded directive

Author: done

:navigation-title: PHP attributes reference

..  include:: /Includes.rst.txt
..  index:: pair: Extbase; PHP attributes
..  _extbase-appendix-attributes:

================================
Extbase PHP attributes reference
================================

..  include:: /ExtensionArchitecture/ExtbaseRewrite/_wip.rst.txt

This page is a quick-reference for all PHP attributes defined and handled by
the Extbase framework itself. It covers :abbr:`ORM (Object Relational Mapping)` attributes (used on model
properties), validation attributes (used on model properties and action
parameters), and controller-level attributes (used on action methods).

It does not cover attributes defined by TYPO3 Core outside of Extbase, or
PHP built-in attributes. Where a more detailed explanation exists in another

> chapter, this page links there.

Reviewer: ```suggestion
It does not cover attributes defined by TYPO3 Core outside of Extbase, or
PHP built-in attributes. 
```
I think the last sentence could be removed. The links already make this obvious, so the sentence feels a bit redundant and slightly interrupts the flow.

Author: done

..  contents:: On this page
    :local:
    :depth: 2

..  _extbase-appendix-attributes-orm:

> ORM (Object Relational Mapping) attributes (persistence)

Reviewer: I miss a short sentence on what ORM is and or a link to more information.

Author: done

=========================================================

These attributes are declared in the :php:`\TYPO3\CMS\Extbase\Attribute\ORM`
namespace and are placed on
model properties to control how Extbase maps them to and from the database.

..  _extbase-appendix-attributes-lazy:

> #[Lazy]

Reviewer: ```suggestion
`#[Lazy]`
```
backtics mark them as code and also make sure the special chars are not interpreted as rst layout

Author: done

-------

:Target: Model property
:Namespace: :php:`\TYPO3\CMS\Extbase\Attribute\ORM\Lazy`
:Parameters: none

> Defers loading of a related object or :php-short:`\TYPO3\CMS\Extbase\Persistence\ObjectStorage` until the property

Reviewer: I would start this section with a simpler explanation of what #[Lazy] does before going into proxies and typed getters. For example:

#[Lazy] tells Extbase to load a related object **only when** it is actually accessed, instead of loading it together with the parent object.

The current first sentence is technically correct, but it is a bit dense as an entry point.

Author: reworded

is first accessed. Without :php:`#[Lazy]`, Extbase loads all related objects
when the parent is loaded — which can trigger N+1 queries in list views.

The proxy resolves itself automatically via PHP magic methods on any property
or method access — you do not need to call :php:`_loadRealInstance()` yourself.

> When applied to a 1:1 relation, the property type must include

Reviewer: or 1:n? same principle?

Author: ObjectStorage resolves this on its own, its also not null, but empty in case of no relations. Only direct relations need this handling.

        #[Validate(validator: 'StringLength', options: ['minimum' => 3, 'maximum' => 50])]
        protected string $title = '';
    }

..  seealso::

    `Built-in validators <https://docs.typo3.org/permalink/extbase-validation-builtin>`_ — all built-in validators and
    their options.

    `Custom validators <https://docs.typo3.org/permalink/extbase-validation-custom>`_ — writing a custom validator.

..  _extbase-appendix-attributes-ignorevalidation:

#[IgnoreValidation]
-------------------

:Target: Action method parameter
:Namespace: :php:`\TYPO3\CMS\Extbase\Attribute\IgnoreValidation`
:Parameters: none (place directly on the parameter)

Suppresses validation for one action parameter. Useful in multi-step forms
where partial state is forwarded between actions:

..  code-block:: php
    :caption: EXT:my_extension/Classes/Controller/ConferenceController.php

    use TYPO3\CMS\Extbase\Attribute\IgnoreValidation;

    class Conference extends AbstractEntity
    {
        public function previewAction(
            #[IgnoreValidation]
            Conference $event
        ): ResponseInterface {
            // validation skipped for $event
        }
    }

> ..  versionchanged:: 14.0

Reviewer: ```suggestion
..  deprecated:: 14.0
```

Author: IgnoreValidation is not deprecated, only the placement at the method level. Putting deprecated here let's it look like the whole attribute will vanish.

Reviewer: Ack

    :php:`#[IgnoreValidation]` must be placed on the **parameter**, not on the
    method with an :php:`argumentName` property. The method-level form was
    deprecated in TYPO3 v14 (Deprecation `#108227
    <https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/14.0/Deprecation-108227-UsageOfIgnoreValidationAndValidateAttributesForParametersAtMethodLevel.html>`__).

..  _extbase-appendix-attributes-controller:

Controller attributes
=====================

These attributes are declared in the :php:`\TYPO3\CMS\Extbase\Attribute`
namespace and are applied to controller action methods to control access and
rate limiting.

.. Full documentation for these attributes is planned for the Controller and Security chapters.

..  _extbase-appendix-attributes-authorize:

#[Authorize]
------------

:Target: Controller action method
:Namespace: :php:`\TYPO3\CMS\Extbase\Attribute\Authorize`

..  versionadded:: 14.0

>     Introduced in Feature `#107826 <https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/14.2/Feature-107826-IntroduceExtbaseActionAuthorizationLogic.html>`__.

Reviewer: Use permalinks as they will not change if the structure or location of the changelog should ever change
```suggestion
    Introduced in Feature `#107826 <https://docs.typo3.org/permalink/changelog:feature-107826-1766220191>`_.
```

Also one underscore is sufficient unless you want to name and reuse links in multiple locations.

Author: done

Declares access requirements for an action method. Extbase checks the
requirements before calling the action and redirects or throws an exception
if they are not met.

.. Full parameter reference and usage examples are coming in the Security chapter.

..  warning::

    :php:`#[Authorize]` handles action-level access (is the user logged in?
    are they in the right group?) but does **not** verify record ownership.
    Always check that a submitted UID belongs to the current user before
    updating or deleting records — Extbase will not do this for you.

..  _extbase-appendix-attributes-ratelimit:

#[RateLimit]
------------

:Target: Controller action method

> :Namespace: :php:`\TYPO3\CMS\Extbase\Attribute\RateLimit`

Reviewer: This is the full FQN, I think?

Author: done

..  versionadded:: 14.0

>     Introduced in Feature `#108982 <https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/14.2/Feature-108982-IntroduceRateLimitingForExtbaseActions.html>`__.

Reviewer: Use permalinks like above

Author: done

This page is a quick-reference for all PHP attributes defined and handled by
the Extbase framework itself. It covers :abbr:`ORM (Object Relational Mapping)` attributes (used on model
properties), validation attributes (used on model properties and action
parameters), and controller-level attributes (used on action methods).

It does not cover attributes defined by TYPO3 Core outside of Extbase, or
PHP built-in attributes.

..  contents:: On this page
    :local:
    :depth: 2

..  _extbase-appendix-attributes-orm:

ORM (Object Relational Mapping) attributes (persistence)
=========================================================

**ORM** (`Object-Relational Mapping <https://en.wikipedia.org/wiki/Object%E2%80%93relational_mapping>`__)
is the mechanism that lets you work with PHP objects instead of raw database
rows. Therefor Extbase maps database fields to Object Model properties and back.
The attributes described below are used to control this process.

The attributes are declared in the :php:`\TYPO3\CMS\Extbase\Attribute\ORM`
namespace and are placed on model properties.

..  _extbase-appendix-attributes-lazy:

`#[Lazy]`
---------

:Target: Model property

> :Full qualified namespace: :php:`\TYPO3\CMS\Extbase\Attribute\ORM\Lazy`

Reviewer 2: ```suggestion
:Fully qualified namespace: :php:`\TYPO3\CMS\Extbase\Attribute\ORM\Lazy`
```

Author: done

:Parameters: none

> Default relation resolving processing loads all related objects together with the parent,

Reviewer 2: ```suggestion
By default, relations are resolved and processed by loading all related objects together with the parent,
```

Author: done

which can harm performance and trigger N+1 queries, for example in list views.
Attributing :php:`#[Lazy]` to a relation of a Model Object will tell Extbase to
load the related object only when it is actively accessed by calling its methods or reading its properties.

> The lazy loading proxy resolves itself automatically via PHP magic methods on any property

Reviewer 2: Why does the example code below have it in the getter then? the part below explains it better IMO and should be merged I think, into a single paragraph?

Author: added a comment to the example.

return type to :php:`?Location`:

..  code-block:: php
    :caption: EXT:my_extension/Classes/Domain/Model/Conference.php

    use TYPO3\CMS\Extbase\Attribute\ORM\Lazy;
    use TYPO3\CMS\Extbase\Persistence\Generic\LazyLoadingProxy;

    class Conference extends AbstractEntity
    {
        #[Lazy]
        protected Location|LazyLoadingProxy|null $location = null;

        public function getLocation(): ?Location
        {
            if ($this->location instanceof LazyLoadingProxy) {
                $this->location = $this->location->_loadRealInstance();
            }
            return $this->location;
        }
    }

If you do not need a narrowly typed getter, the check is unnecessary — accessing
the proxy in any way triggers resolution automatically.

..  seealso::

    `Relations and ObjectStorage <https://docs.typo3.org/permalink/extbase-domain-model-relations>`_ — relations and ObjectStorage
    on the model page.

    `Persistence relations <https://docs.typo3.org/permalink/extbase-persistence-relations>`_ — full lazy loading reference
    and the N+1 query trap.

..  _extbase-appendix-attributes-cascade:

`#[Cascade]`
------------

:Target: Model property

> :Full qualified namespace: :php:`\TYPO3\CMS\Extbase\Attribute\ORM\Cascade`

Reviewer 2: (Use "fully" everywhere)

Author: done

:Parameters: :php:`string|null $value = null`

> Controls what happens to related objects when the owning object is deleted.

Reviewer 2: ```suggestion
Controls what happens to related objects when the parent object is deleted.
```

Author: done

Currently only :php:`'remove'` is supported:

..  code-block:: php
    :caption: EXT:my_extension/Classes/Domain/Model/Conference.php

    use TYPO3\CMS\Extbase\Attribute\ORM\Cascade;
    use TYPO3\CMS\Extbase\Attribute\ORM\Lazy;

    class Conference extends AbstractEntity
    {
        #[Lazy]
        #[Cascade('remove')]
        protected ObjectStorage $comments;
    }

Without :php:`#[Cascade('remove')]`, deleting the parent object leaves related
records in the database as orphans. With it, Extbase deletes the related
objects automatically via the repository.

..  note::

    Only :php:`'remove'` is supported. Other Doctrine cascade operations

>     (:php:`'persist'`, :php:`'merge'`, etc.) are not implemented in Extbase.

Reviewer 2: Add a note that this cascading only applies to deleting objects via repository methods. Deleting an object in the backend will only also cascade if configured in TCA, right?

Author: done

..  _extbase-appendix-attributes-transient:

> `#[Transient]`

Reviewer 2: IIRC there was some noise about if this actually works in the core and/or should be removed, did you by chance research this if it works atm?

Author: codewise confirmed, it works. If not, it is still supposed to work and any discussions otherwise belong to forge.

    deprecated in TYPO3 v14 (Deprecation `#108227
    <https://docs.typo3.org/c/typo3/cms-core/main/en-us/Changelog/14.0/Deprecation-108227-UsageOfIgnoreValidationAndValidateAttributesForParametersAtMethodLevel.html>`__).

..  _extbase-appendix-attributes-controller:

Controller attributes
=====================

These attributes are declared in the :php:`\TYPO3\CMS\Extbase\Attribute`
namespace and are applied to controller action methods to control access and
rate limiting.

.. Full documentation for these attributes is planned for the Controller and Security chapters.

..  _extbase-appendix-attributes-authorize:

`#[Authorize]`
--------------

:Target: Controller action method
:Full qualified namespace: :php:`\TYPO3\CMS\Extbase\Attribute\Authorize`

..  versionadded:: 14.0

    Introduced in Feature `#107826 <https://docs.typo3.org/permalink/changelog:feature-107826-1766220191>`_.

Declares access requirements for an action method. Extbase checks the
requirements before calling the action and redirects or throws an exception
if they are not met.

.. Full parameter reference and usage examples are coming in the Security chapter.

..  warning::

>     :php:`#[Authorize]` handles action-level access (is the user logged in?

Reviewer 2: ```suggestion
    :php:`#[Authorize]` handles action-level access ("is the user logged in?
```

Author: done

>     are they in the right group?) but does **not** verify record ownership.

Reviewer 2: ```suggestion
    are they in the right group?") but does **not** verify record ownership.
```

Author: done
